Add a way to convert std ranges to wtf containers

colelao opened the pull request:
#### aa649b89f25e6ca725769ed3300e2264fbee3efb
<pre>
Add a way to convert std ranges to wtf containers
<a href="https://bugs.webkit.org/show_bug.cgi?id=303699">https://bugs.webkit.org/show_bug.cgi?id=303699</a>
<a href="https://rdar.apple.com/165990982">rdar://165990982</a>

Reviewed by Sam Weinig.

This patch exposes a functor WTF::rangesTo&lt;ContainerT&gt; that will convert
any range to any WTF container that implements a constructor that takes
a std::from_range_t and an std::ranges::input_range.

I added the proper constructors to the wtf Vector and HashMap. The constructors
did not need use the disambiguation tag (std::from_range_t), but I did this
so it would not interfere with any existing constructors.

After this patch the following is possible:

Vector&lt;int&gt; v = std::views::iota(0, 10) | WTF::rangeTo&lt;Vector&lt;int&gt;&gt;;

Tests: Tools/TestWebKitAPI/Tests/WTF/HashMap.cpp
       Tools/TestWebKitAPI/Tests/WTF/Vector.cpp

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/HashMap.h:
* Source/WTF/wtf/RangeAdaptors.h: Added.
(WTF::RangeAdaptorClosure::operator|):
(WTF::FromRangeConverter::requires):
(WTF::FromRangeConverter::operator() const):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
* Tools/TestWebKitAPI/Tests/WTF/HashMap.cpp:
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorBasicTypes)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorInitializerList)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorEmptyRange)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorMoveOnlyValues)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorStringKeys)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorCustomHashTraits)):
(TestWebKitAPI::TEST(WTF_HashMap, RangeToBasicUsage)):
(TestWebKitAPI::TEST(WTF_HashMap, RangeToChainedOperations)):
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorBasicTypes)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorArray)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorInitializerList)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorEmptyRange)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorStrings)):
(TestWebKitAPI::TEST(WTF_Vector, RangeToBasicUsage)):
(TestWebKitAPI::TEST(WTF_Vector, RangeToChainedOperations)):
(TestWebKitAPI::TEST(WTF_Vector, RangeToMoveOnly)):

Canonical link: <a href="https://commits.webkit.org/304212@main">https://commits.webkit.org/304212@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/3e08a69f59ec125781c025e8c974cd49881053f9

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows | Apple Internal |
| ----- | ---------------------- | ------- |  ----- |  --------- | ------ |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/134857 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/159/builds/7286 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/46103 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/142370 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/86742 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 ios-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/22569445-9297-4fcd-b342-0ab3b58e44ff) 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/136726 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/155/builds/7901 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/156/builds/7133 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/103050 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/70336 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/695329f7-055d-4dd9-a0ee-b9d1cbea784f) 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/137803 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/162/builds/5585 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/120879 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/83890 "Passed tests") | | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 vision-apple](https://ews-bridge.webkit.apple.com/builds/sw/T-276/f2c6d2ed-bd8b-4d70-a499-64172ae81bd0/6faf4d20-d760-4271-8410-6107d6c883af) 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/154/builds/5414 "Passed tests") | [✅ 🧪 api-mac-debug](https://ews-build.webkit.org/#/builders/165/builds/3026 "Passed tests") | [✅ 🛠 wpe-cairo-libwebrtc](https://ews-build.webkit.org/#/builders/166/builds/2963 "Built successfully") | | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/126886 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/114629 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/39031 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/145067 "Built successfully") | | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/133359 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/153/builds/6961 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/39607 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/111433 "Passed tests") | | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/160/builds/7020 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/161/builds/5842 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/111776 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/28369 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/164/builds/5249 "Passed tests") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/117159 "Passed tests") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/60891 "Built successfully") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/158/builds/7008 "Built successfully") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/35334 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/166256 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/157/builds/6792 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/70586 "Built successfully") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/43454 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/163/builds/7028 "Built successfully") | | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/152/builds/6902 "Built successfully") | | | | 
<!--EWS-Status-Bubble-End-->

nullhook commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,68 @@
+
+namespace WTF {
+
+struct FromRange {
+    explicit FromRange() = default;
+};
+
+// Global instance for use as tag
+inline constexpr FromRange fromRange { };
```
is there a reason we want to use this over a similar tag type `std::in_place_t` to disambiguate? other than semantics.

colelao replied:
This new tag is here for this specific use case. It makes sure that there is no clobbering of existing constructors, and that any WTF container that wants to be convertible from a range has to use this tag in their constructor to handle the conversion.

nullhook replied:
right, and is the following effectively the same but semantically different from the `FromRange` tag type? (i'm not suggesting a change, btw. just curious)

```c++
template <typename Container>
struct FromRangeConverter : RangeAdaptorClosure<FromRangeConverter<Container>> {
    template <std::ranges::input_range Range>
        requires requires(Range&& range) { Container(std::in_place_type_t<Container> { }, std::forward<Range>(range)); }
     ...
}

template <typename Range>
requires std::ranges::input_range<Range> && std::convertible_to<std::ranges::range_value_t<Range>, T>
    explicit Vector(std::in_place_type_t<T>, Range&& range) { ... }
```

colelao replied:
Yes, that's doing the same thing. The only difference is the "social contract" of what the tag will do.

weinig replied:
I think using the new tag makes sense. This follows from how the standard works. 

At some point, I hope we can use the standard one (https://en.cppreference.com/w/cpp/ranges/from_range.html)?

weinig replied:
I would probably leave a comment and a link like the one above, explaining the intent. Either that we are doing this because we don't want to use the standard one, or because we can't use it yet.

nullhook commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,68 @@
+template <typename Container>
+struct FromRangeConverter : RangeAdaptorClosure<FromRangeConverter<Container>> {
+    template <std::ranges::input_range Range>
+        requires requires(Range&& range) { Container(fromRange, std::forward<Range>(range)); }
```
aside: something i've noticed that global objects `fromRange` has a storage cost. not sure if i'm mimicking the right setup here:

https://godbolt.org/z/8fMeTdWxb

```
fromRange:
        .zero   1
```

rather, if you throw a temporary, that cost is gone

```cpp
foo.add<int>(FromRange { }, a, b);
```

obviously, all of this is optimized when compiled with opt flags.

colelao replied:
Yes, there is a small storage cost when compiled without optimizations, but it becomes sugar with any optimization flags.

cdumez commented on Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
```diff
@@ -2140,4 +2140,98 @@ TEST(WTF_Vector, InsertFill)
+TEST(WTF_Vector, fromRangeConstructorBasicTypes)
+{
+    std::vector<int> stdVec { 1, 2, 3, 4, 5 };
+    Vector<int> wtfVec(fromRange, stdVec);
```
I mean, would this just work?
```
Vector<int> wtfVec(stdVec);
```

By implicitly converting the stdVec to a std::span? Maybe we'd need the `std::span { }` but still not too bad. Also, we don't usually have to deal with std containers much in WebKit so it is not an obvious use case to me.

colelao replied:
The idea is that this would give us the power to convert anything that conforms to the `std::ranges::input_range` concept into any WTF container that implements a constructor with the `fromRange` tag. 

This would also allow us to use the std::ranges library functions to do things like this:
```
Vector<int> doSomething(const Vector<int>& vec)
{
    return vec
        | std::views::filter([](int n) { return (n % 2); })
        | std::views::transform([](int n) { return n * 2; })
        | std::views::reverse()
        | WTF::rangeTo<Vector<int>>(); // convert the std::range view to a WTF Vector
}
```

There is an `std::ranges::to` function, but our containers do not conform to the `std::ranges::output_range` concept, so I thought this was a safe way to add that functionality.

geoffreygaren replied:
I guess I understand the incremental development goal in the abstract. But it's jarring to see it introduced using std::vector. Makes it feel like it's not a good fit if it can't work with WTF::Vector.

Is your plan to change these tests to use WTF::Vector in a follow-up patch?

colelao replied:
It does work with WTF::Vector. I thought it might be useful to test the WTF::rangeTo with an input range that is known to be stable and separate from WTF. Although, the most likely use case is from one WTF container to another WTF container. 

I will change the tests to reflect this.

weinig commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -94,6 +95,20 @@ class HashMap final {
     HashMap() = default;
 
+    template <typename Range>
+    requires std::ranges::input_range<Range> && std::convertible_to<std::ranges::range_value_t<Range>, KeyValuePairType>
```
Please indent this 4 more spaces..

weinig commented on Source/WTF/wtf/HashMap.h:
```diff
@@ -94,6 +95,20 @@ class HashMap final {
 public:
     HashMap() = default;
 
+    template <typename Range>
```
Our current style guide has no space after `template`.
```suggestion
    template<typename Range>
```

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+// Global instance for use as tag
+inline constexpr FromRange fromRange { };
+
+template <typename Derived>
```
```suggestion
template<typename Derived>
```

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+
+template <typename Derived>
+struct RangeAdaptorClosure {
+    template <std::ranges::input_range Range>
```
```suggestion
    template<std::ranges::input_range Range>
```

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+inline constexpr FromRange fromRange { };
+
+template <typename Derived>
+struct RangeAdaptorClosure {
```
Similarly, please leave a comment about how this relates to std::range_adaptor_closure (https://en.cppreference.com/w/cpp/ranges/range_adaptor_closure.html).

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+        return adaptor(std::forward<Range>(range));
+    }
+
+    template <typename OtherClosure>
```
```suggestion
    template<typename OtherClosure>
```

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+    }
+};
+
+template <typename Container>
```
```suggestion
template<typename Container>
```

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+
+template <typename Container>
+struct FromRangeConverter : RangeAdaptorClosure<FromRangeConverter<Container>> {
+    template <std::ranges::input_range Range>
```
```suggestion
    template<std::ranges::input_range Range>
```

weinig commented on Source/WTF/wtf/RangeAdaptors.h:
```diff
@@ -0,0 +1,71 @@
+    }
+};
+
+template <typename Container>
```
```suggestion
template<typename Container>
```

weinig commented on Source/WTF/wtf/Vector.h:
```diff
@@ -602,6 +604,20 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     {
     }
 
+    template <typename Range>
```
```suggestion
    template<typename Range>
```

weinig commented on Source/WTF/wtf/Vector.h:
```diff
@@ -602,6 +604,20 @@ class Vector : private VectorBuffer<T, inlineCapacity, Malloc> {
     }
 
+    template <typename Range>
+    requires std::ranges::input_range<Range> && std::convertible_to<std::ranges::range_value_t<Range>, T>
```
Please indent this 4 more spaces.

weinig reviewed (approved):
Very nice.

darinadler reviewed (commented):
I always enjoy when our WTF tool kit gets some helpful additions. One thing I usually prefer is that we find a few places where it makes WebKit code better, rather than adding it first and then using it later. I’m a little disappointed to not see, in this pull request, one or two places in JavaScriptCore, WebCore, or WebKit where we adopt this that show how well it pays off.

And I would like to second the comments about explaining any time we make an WTF equivalent to something in the standard library. We want to say whether this is "better" in some way or "just until we get to C++23" or whatever our rationale is. We’ve benefitted from not creating our entire separate universe when possible.

colelao committed:
Add a way to convert std ranges to wtf containers
https://bugs.webkit.org/show_bug.cgi?id=303699
rdar://165990982

Reviewed by Sam Weinig.

This patch exposes a functor WTF::rangesTo<ContainerT> that will convert
any range to any WTF container that implements a constructor that takes
a std::from_range_t and an std::ranges::input_range.

I added the proper constructors to the wtf Vector and HashMap. The constructors
did not need use the disambiguation tag (std::from_range_t), but I did this
so it would not interfere with any existing constructors.

After this patch the following is possible:

Vector<int> v = std::views::iota(0, 10) | WTF::rangeTo<Vector<int>>;

Tests: Tools/TestWebKitAPI/Tests/WTF/HashMap.cpp
       Tools/TestWebKitAPI/Tests/WTF/Vector.cpp

* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/HashMap.h:
* Source/WTF/wtf/RangeAdaptors.h: Added.
(WTF::RangeAdaptorClosure::operator|):
(WTF::FromRangeConverter::requires):
(WTF::FromRangeConverter::operator() const):
* Source/WTF/wtf/Vector.h:
(WTF::Vector::Vector):
* Tools/TestWebKitAPI/Tests/WTF/HashMap.cpp:
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorBasicTypes)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorInitializerList)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorEmptyRange)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorMoveOnlyValues)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorStringKeys)):
(TestWebKitAPI::TEST(WTF_HashMap, fromRangeConstructorCustomHashTraits)):
(TestWebKitAPI::TEST(WTF_HashMap, RangeToBasicUsage)):
(TestWebKitAPI::TEST(WTF_HashMap, RangeToChainedOperations)):
* Tools/TestWebKitAPI/Tests/WTF/Vector.cpp:
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorBasicTypes)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorArray)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorInitializerList)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorEmptyRange)):
(TestWebKitAPI::TEST(WTF_Vector, fromRangeConstructorStrings)):
(TestWebKitAPI::TEST(WTF_Vector, RangeToBasicUsage)):
(TestWebKitAPI::TEST(WTF_Vector, RangeToChainedOperations)):
(TestWebKitAPI::TEST(WTF_Vector, RangeToMoveOnly)):

Canonical link: https://commits.webkit.org/304212@main
